source/agora/consensus/data/UTXOSet.d

```diff
@@ -81,25 +126,105 @@ public class UTXOSet
 
         Params:
             tx = the transaction
+            height = Height of the block where UTXO will be unlocked
 
     ***************************************************************************/
 
-    public void updateUTXOCache (const ref Transaction tx) @safe
+    public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
     {
-        foreach (const ref input; tx.inputs)
+        if (tx.type == TxType.Freeze)
         {
-            auto utxo_hash = getHash(input.previous, input.index);
-            this.utxo_db.remove(utxo_hash);
+            foreach (const ref input; tx.inputs)
+            {
+                auto utxo_hash = getHash(input.previous, input.index);
+                this.utxo_db.remove(utxo_hash);
+            }
+
+            ulong unlock_height = height + 1;
+
+            Hash tx_hash = tx.hashFull();
+            foreach (idx, output; tx.outputs)
+            {
+                auto utxo_hash = getHash(tx_hash, idx);
+                const UTXOSetValue value = UTXOSetValue(unlock_height, tx.type, output);
+                this.utxo_db[utxo_hash] = value;
+            }
         }
-
-        Hash tx_hash = tx.hashFull();
-        foreach (idx, output; tx.outputs)
+        else
         {
-            auto utxo_hash = getHash(tx_hash, idx);
-            this.utxo_db[utxo_hash] = output;
+            TxType tx_type = TxType.Payment;
+            if (tx.inputs.length > 0)
+            {
+                UTXOSetValue utxo_value;
+                int n_payment = 0;
+                int n_freeze = 0;
+                foreach (const ref input; tx.inputs)
+                {
+                    if (this.getUTXOSetValue(input.previous, input.index, utxo_value))
+                    {
+                        if (utxo_value.type == TxType.Payment)
+                            n_payment++;
+                        else
+                            n_freeze++;
+                    }
+                    else if (input.previous == Hash.init)
+                        n_payment++;
+                }
+
+                if (n_freeze == tx.inputs.length)
```

Comment:
This is the part I find a bit confusing. The new outputs (UTXOs) will be be melting only if **all** of the inputs are frozen UTXOs. But if one of the was a Payment, then all of the outputs will not be melting, they will be instantly melted?